### ISSUE TITLE
Ain't no sound match for the incomplete

### ISSUE THREAD

**Issue by dhil** [links-lang/links#586]
As of writing we do not support exhaustiveness checking for pattern, meaning the following program won't produce a compile-time error or warning
```links
fun f(xs) {
  switch (xs) {
     case [] -> ()
  }
}
```
As a consequence the application `f([42])` causes a runtime error. From a certain point of view that is the expected behaviour. However, here comes the point: the above program has inferred type
```links
f : ([a]) -> ()
```
which is a *tame* type. In fact it can even be given the stricter type
```
sig f : ([a]) {}-> ()
```
meaning that it is *pure*.

* Since the function is tame, it can be used in database queries to generate well-formed results, except it can't.
* Since the function is pure, running it will cause no observable effects, except it won't.

Arguably, an non-exhaustive `switch` should introduce the `wild` effect into the context such that the inferred type of the above function would be
```
f : ([a]) ~> ()
```

To achieve this we need to perform an exhaustiveness check on patterns during type checking.

**Comment by dhil** [comment]
Related to #37.

**Comment by jamescheney** [comment]
Well, we also aren't checking that datatypes used in tame code are strictly positive, which means that recursive functions 
Fixing #37 (and #22) would be good in any case though.  #37 is really about lists being a special built-in type rather than a plain Links datatype.

**Comment by dhil** [comment]
> Well, we also aren't checking that datatypes used in tame code are strictly positive, which means that recursive functions

Yes, that is another problem, which means one would be able to write a non-terminating query.

**Comment by jamescheney** [comment]
Thinking about this and #37, #47, I think the underlying problem is that in Links, the built-in list type is punned in an unhelpful way: on the server/client it stands for Cons/Nil lists (which we could imagine defining as a user-defined recursive variant type as proposed in #47), but on the database it stands for multisets, for which pattern matching doesn't necessarily make sense.  

(Caveat: using things like limit/offset and the fact that SQL lets you silently coerce a singleton set to an element,  it is possible to emulate some kinds of pattern matching on lists, but I don't think it's a good idea or something we want to encourage).

Because the list type is built-in, any pattern matching against either Cons, Nil or a list pattern is automatically promoted to having the full list type, laundering away any more specific information about the patterns matched.  If we had a user-defined list type (separate from the collection type used for database queries) then pattern matching would be typechecked using variant types, so the example above would still typecheck and have a tame type, but the type would be 
```
f : ([|Nil|]) -> ()
```
so, would only be applicable to a (user defined) empty list, not to an arbitrary list.  

I am inclined to resolve this and #37 by inspecting the patterns during typechecking and requiring a wild context if a switch case contains any subpatterns matching against Nil, Cons, or List patterns.  This should also help make the IR generated from patterns more likely to  typecheck; I expect that this is not the case now because list pattern matching currently desugars to calls to hd and tail, leading to wild effects where there weren't necessarily before.

**Comment by jamescheney** [comment]
One further refinement: it's only `Cons` and list patterns (that could be desugared to one or more Cons patterns) that should incur a wild effect.  Pattern matching against `[]` / `Nil` is fine - this can be done in the database using the emptiness test.  (Indeed, in the prelude, the `empty` function is defined by pattern matching on `[]`!)

**Comment by dhil** [comment]
This particular issue is more general than the built-in list datatype. Consider the following program
```links
typename T = [|Foo|Bar|];

sig f : ((T, T)) {}-> () 
fun f(p) { 
   switch (p) { case (Foo, Foo) -> () case (Bar, Bar) -> () } 
}
```
The pattern match is incomplete, yet the type of `f` is pure. Invoking `f((Foo,Bar))` or `f((Bar,Foo))` cause a runtime error.

**Comment by jamescheney** [comment]
Good point.  I think @frank-emrich has been thinking about ways to typecheck such pattern matches more precisely...

I think pattern matching against Cons should not be allowed in queries no matter what.  If we had an exhaustiveness checker for variant patterns, we could give non-exhaustive patterns the wild effect to prevent this.  I'm not really sure how useful it is to allow variant pattern matching inside queries - if we just made any use of "switch" have the wild effect, then that would rule out this problem.  On the other hand, being able to use option types to deal with nulls inside queries might be handy.
